core/src/builtin/types.rs

```diff
@@ -1,15 +1,30 @@
 use crate::Printable;
 use crate::{Attr, ContextRef, Ty, TyAssembly, Type};
 use std::collections::HashMap;
-use tir_macros::dialect_type;
+use tir_macros::{dialect_type, dialect_type_with_extensions};
 
 use crate as tir_core;
 
 use crate::builtin::DIALECT_NAME;
 
-dialect_type!(FuncType);
+dialect_type_with_extensions!(FuncType);
 dialect_type!(VoidType);
-dialect_type!(IntType);
+dialect_type_with_extensions!(IntType);
+
+impl TyAssembly for VoidType {
+    fn print_assembly(
+        _attrs: &HashMap<String, tir_core::Attr>,
+        fmt: &mut dyn tir_core::IRFormatter,
+    ) {
+        fmt.write_direct("void");
+    }
+
+    fn parse_assembly(
+        input: &mut tir_core::parser::ParseStream<'_>,
+    ) -> tir_core::parser::AsmPResult<std::collections::HashMap<String, tir_core::Attr>> {
+        tir_core::parser::skip_attrs(input)
```

Comment:
Is it enough? No need to explicitly parse void?

Reply:
Right, it's parsed by caller